### PR TITLE
make non-PartialEq-typed consts as patterns a hard error

### LINKED ISSUE #116122: Tracking issue for `const_patterns_without_partial_eq` lint
We used to (accidentally) accept some constants in patterns that don't even implement `PartialEq`. This is an accident and we'd like to hard-reject all such cases in the future. This lint catches those cases.

Relevant PRs:
- https://github.com/rust-lang/rust/pull/115893: lint implementation

### PR DESCRIPTION
**Pull request by RalfJung**
This lint was introduced in https://github.com/rust-lang/rust/pull/115893, for Rust 1.74, so we just had the third stable release where this is shown as a future-compat lint (which is shown for dependencies). Not a single comment or backreference showed up in the tracking issue, https://github.com/rust-lang/rust/issues/116122. So this seems fairly safe to turn into a hard error.

Of course we should do a crater run first.

This is part of https://github.com/rust-lang/rust/issues/120362.
Closes https://github.com/rust-lang/rust/issues/116122.

### CODE REVIEW

**Comment by fmease** [comment]
Both root regressions look … spurious?

**Comment by RalfJung** [comment]
Yes indeed.

Nominating for t-lang: this is the next step for the const pattern RFC. It makes constants of non-`PartialEq` type in patterns a hard error. This has been shown in future-breakage reports since Rust 1.74. Crater found no regressions.

**Comment by RalfJung** [comment]
Turns out the check currently misses some cases:
https://github.com/rust-lang/rust/issues/121007

However, those are extremely hard to even trigger -- even harder than triggering the lint. Not sure if that should block this PR. Also I don't see how we can fix #121007 in a way that triggers a lint rather than a hard error; the most obvious fix is to insert a call to `PartialEq::eq` in dead code but then we'd obviously get a hard error in case the bound is not satisfied.

**Review by oli-obk — APPROVED** [review]

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by pnkfelix** [comment]
🎉 

@rfcbot reviewed

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

As evidenced by the proposed FCP merge and the fact that this is now in FCP, we discussed this in the lang triage call today and seemed to be in favor of doing this.

Thanks to @RalfJung for working to push this forward.

**Comment by fmease** [comment]
merge conflicts 
@rustbot author

**Comment by RalfJung** [comment]
@oli-obk you marked this as "approved", I assume that's an "r=me after FCP passes".
@bors r=oli-obk

**Pull request merged**
